keps/sig-node/3066-subsecond-probes/README.md

```diff
@@ -0,0 +1,645 @@
+	// Must be 1 for liveness and startup.
+	// +optional
+	SuccessThreshold int32
+	// Minimum consecutive failures for the probe to be considered failed after having succeeded.
+	// +optional
+	FailureThreshold int32
+	// Optional duration in seconds the pod needs to terminate gracefully upon probe failure.
+	// The grace period is the duration in seconds after the processes running in the pod are sent
+	// a termination signal and the time when the processes are forcibly halted with a kill signal.
+	// Set this value longer than the expected cleanup time for your process.
+	// If this value is nil, the pod's terminationGracePeriodSeconds will be used. Otherwise, this
+	// value overrides the value provided by the pod spec.
+	// Value must be non-negative integer. The value zero indicates stop immediately via
+	// the kill signal (no opportunity to shut down).
+	// This is a beta field and requires enabling ProbeTerminationGracePeriod feature gate.
+	// +optional
+	TerminationGracePeriodSeconds *int64
+}
+```
+
+### Existing Behavior
+
+Taking the Seconds fields, in order of the struct.
+InitialDelaySeconds
+TimeoutSeconds
+PeriodSeconds
+
+#### Defaulting logic
+ - InitialDelaySeconds, has no defaulting logic.
+   Therefore it would get the golang defaulting logic, and default to 0.
+ - TimeoutSeconds, defaults to 1 second if unset (which includes the explicitly set to 0 state). This is the last ending time before the timeout fails. Defaulting, thus, results in slow failures (when one considers a full second of time in cpu time).
+https://github.com/kubernetes/kubernetes/blob/e19964183377d0ec2052d1f1fa930c4d7575bd50/pkg/apis/core/v1/defaults.go#L224-L226
+
+```go
+ 	if obj.TimeoutSeconds == 0 {
+		obj.TimeoutSeconds = 1
+	}
+```
+
+ - Period Seconds is the biggest hurdle in charging and overall cost to the cluster, but also the most useful if set correctly.
+PeriodSeconds defaults to 10 seconds if unset (which includes the explicitly set to 0 state). Fast failure detections are thus blocked with existing defaults.
+https://github.com/kubernetes/kubernetes/blob/e19964183377d0ec2052d1f1fa930c4d7575bd50/pkg/apis/core/v1/defaults.go#L227-L229
+
+```go
+	if obj.PeriodSeconds == 0 {
+		obj.PeriodSeconds = 10
+	}
+```
+
+#### Validation of fields
+
+Must be non-negative, Zero or greater.
+```go
+	allErrs = append(allErrs, ValidateNonnegativeField(int64(probe.InitialDelaySeconds), fldPath.Child("initialDelaySeconds"))...)
+	allErrs = append(allErrs, ValidateNonnegativeField(int64(probe.TimeoutSeconds), fldPath.Child("timeoutSeconds"))...)
+	allErrs = append(allErrs, ValidateNonnegativeField(int64(probe.PeriodSeconds), fldPath.Child("periodSeconds"))...)
+
+```
+
+#### Fields to Add
+What fields may be necessary to add?
+
+In pkg/apis/core/types.go
+
+```go
+
+type Probe struct {
+    ...
+    ...
+    // How often (in milliseconds) to offset PeriodSeconds when performing the probe.
+	// +optional
+	PeriodMilliseconds *int32
```

Comment:
Will this be applied to every iteration or just the first? 
If to the first, maybe the name should be more expressive - `FirstPeriodOffsetMilliseconds` ? TBH, I am a bit concerned how intuitive this is to end users.

Reply:
nod will document in more detail on the api and field comment.. idea is shorter period until success longer period would continue... 1.5 period... would continue .5 period would continue till success as suggested.. .5, 1, 1.5..success 2.5, 3.5...